<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Gerhard Schöfl" />

<meta name="date" content="2020-10-08" />

<title>Introduction to reutils</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction to reutils</h1>
<h4 class="author">Gerhard Schöfl</h4>
<h4 class="date">2020-10-08</h4>



<p><code>reutils</code> is an R package for interfacing with NCBI databases such as PubMed, Genbank, or GEO via the Entrez Programming Utilities (<a href="https://www.ncbi.nlm.nih.gov/books/NBK25501/">EUtils</a>). It provides access to the nine basic <em>eutils</em>: <code>einfo</code>, <code>esearch</code>, <code>esummary</code>, <code>epost</code>, <code>efetch</code>, <code>elink</code>, <code>egquery</code>, <code>espell</code>, and <code>ecitmatch</code>.</p>
<p>Please check the relevant <a href="https://www.ncbi.nlm.nih.gov/books/NBK25497/#chapter2.Usage_Guidelines_and_Requiremen">usage guidelines</a> when using these services. Note that Entrez server requests are subject to frequency limits. Consider obtaining an NCBI API key if are a heavy user of E-utilities.</p>
<div id="important-functions" class="section level2">
<h2>Important functions</h2>
<p>With nine E-Utilities, NCBI provides a programmatical interface to the Entrez query and database system for searching and retrieving requested data</p>
<p>Each of these tools corresponds to an <code>R</code> function in the reutils package described below.</p>
<div id="esearch" class="section level4">
<h4><code>esearch</code></h4>
<p><code>esearch</code>: search and retrieve a list of primary UIDs or the NCBI History Server information (queryKey and webEnv). The objects returned by <code>esearch</code> can be passed on directly to <code>epost</code>, <code>esummary</code>, <code>elink</code>, or <code>efetch</code>.</p>
</div>
<div id="efetch" class="section level4">
<h4><code>efetch</code></h4>
<p><code>efetch</code>: retrieve data records from NCBI in a specified retrieval type and retrieval mode as given in this <a href="https://www.ncbi.nlm.nih.gov/books/NBK25499/table/chapter4.T._valid_values_of__retmode_and/?report=objectonly">table</a>. Data are returned as XML or text documents.</p>
</div>
<div id="esummary" class="section level4">
<h4><code>esummary</code></h4>
<p><code>esummary</code>: retrieve Entrez database summaries (DocSums) from a list of primary UIDs (Provided as a character vector or as an <code>esearch</code> object)</p>
</div>
<div id="elink" class="section level4">
<h4><code>elink</code></h4>
<p><code>elink</code>: retrieve a list of UIDs (and relevancy scores) from a target database that are related to a set of UIDs provided by the user. The objects returned by <code>elink</code> can be passed on directly to <code>epost</code>, <code>esummary</code>, or <code>efetch</code>.</p>
</div>
<div id="einfo" class="section level4">
<h4><code>einfo</code></h4>
<p><code>einfo</code>: provide field names, term counts, last update, and available updates for each database.</p>
</div>
<div id="epost" class="section level4">
<h4><code>epost</code></h4>
<p><code>epost</code>: upload primary UIDs to the users’s Web Environment on the Entrez history server for subsequent use with <code>esummary</code>, <code>elink</code>, or <code>efetch</code>.</p>
</div>
</div>
<div id="examples" class="section level2">
<h2>Examples</h2>
<div id="esearch-searching-the-entrez-databases" class="section level3">
<h3><code>esearch</code>: Searching the Entrez databases</h3>
<p>Let’s search PubMed for articles with Chlamydia psittaci in the title that have been published in 2020 and retrieve a list of PubMed IDs (PMIDs).</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>pmid &lt;-<span class="st"> </span><span class="kw">esearch</span>(<span class="st">&quot;Chlamydia psittaci[titl] and 2020[pdat]&quot;</span>, <span class="st">&quot;pubmed&quot;</span>)</span>
<span id="cb1-2"><a href="#cb1-2"></a>pmid</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="co">#&gt; Object of class &#39;esearch&#39; </span></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="co">#&gt; List of UIDs from the &#39;pubmed&#39; database.</span></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="co">#&gt;  [1] &quot;32848009&quot; &quot;32830314&quot; &quot;32416138&quot; &quot;32326284&quot; &quot;32316620&quot; &quot;32314307&quot;</span></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="co">#&gt;  [7] &quot;32290117&quot; &quot;32183481&quot; &quot;32178660&quot; &quot;32135200&quot; &quot;32071972&quot; &quot;32057555&quot;</span></span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="co">#&gt; [13] &quot;32050885&quot; &quot;31951466&quot; &quot;31910921&quot; &quot;31755895&quot; &quot;31436332&quot;</span></span></code></pre></div>
<p>Alternatively we can collect the PMIDs on the history server.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>pmid2 &lt;-<span class="st"> </span><span class="kw">esearch</span>(<span class="st">&quot;Chlamydia psittaci[titl] and 2020[pdat]&quot;</span>, <span class="st">&quot;pubmed&quot;</span>, <span class="dt">usehistory =</span> <span class="ot">TRUE</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a>pmid2</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="co">#&gt; Object of class &#39;esearch&#39; </span></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="co">#&gt; Web Environment for the &#39;pubmed&#39; database.</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="co">#&gt; Number of UIDs stored on the History server: 17</span></span>
<span id="cb2-6"><a href="#cb2-6"></a><span class="co">#&gt; Query Key: 1</span></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co">#&gt; WebEnv: MCID_5f7f425c6377c3335612d79c</span></span></code></pre></div>
<p>We can also use <code>esearch</code> to search GenBank. Here we do a search for polymorphic membrane proteins (PMPs) in Chlamydiaceae.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>cpaf &lt;-<span class="st"> </span><span class="kw">esearch</span>(<span class="st">&quot;Chlamydiaceae[orgn] and PMP[gene]&quot;</span>, <span class="st">&quot;nucleotide&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2"></a>cpaf</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="co">#&gt; Object of class &#39;esearch&#39; </span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co">#&gt; List of UIDs from the &#39;nucleotide&#39; database.</span></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co">#&gt; [1] &quot;519865230&quot; &quot;410810883&quot; &quot;313847556&quot; &quot;532821947&quot; &quot;519796743&quot; &quot;532821218&quot;</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co">#&gt; [7] &quot;519794601&quot;</span></span></code></pre></div>
<p>Some accessors for <code>esearch</code> objects</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">getUrl</span>(cpaf)</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co">#&gt; [1] &quot;https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?term=Chlamydiaceae%5Borgn%5D%20AND%20PMP%5Bgene%5D&amp;db=nucleotide&amp;retstart=0&amp;retmax=100&amp;rettype=uilist&amp;retmode=xml&amp;email=gerhard.schofl%40gmail.com&amp;tool=reutils&quot;</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw">getError</span>(cpaf)</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="co">#&gt; No errors</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="kw">database</span>(cpaf)</span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="co">#&gt; [1] &quot;nucleotide&quot;</span></span></code></pre></div>
<p>Extract a vector of GIs:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">uid</span>(cpaf)</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="co">#&gt; [1] &quot;519865230&quot; &quot;410810883&quot; &quot;313847556&quot; &quot;532821947&quot; &quot;519796743&quot; &quot;532821218&quot;</span></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="co">#&gt; [7] &quot;519794601&quot;</span></span></code></pre></div>
<p>Get query key and web environment:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw">querykey</span>(pmid2)</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="co">#&gt; [1] 1</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="kw">webenv</span>(pmid2)</span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="co">#&gt; [1] &quot;MCID_5f7f425c6377c3335612d79c&quot;</span></span></code></pre></div>
<p>Extract the content of an EUtil request as XML.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw">content</span>(cpaf, <span class="st">&quot;xml&quot;</span>)</span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="co">#&gt; &lt;?xml version=&quot;1.0&quot; encoding=&quot;UTF-8&quot;?&gt;</span></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="co">#&gt; &lt;!DOCTYPE eSearchResult PUBLIC &quot;-//NLM//DTD esearch 20060628//EN&quot; &quot;https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20060628/esearch.dtd&quot;&gt;</span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="co">#&gt; &lt;eSearchResult&gt;</span></span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="co">#&gt;   &lt;Count&gt;7&lt;/Count&gt;</span></span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="co">#&gt;   &lt;RetMax&gt;7&lt;/RetMax&gt;</span></span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="co">#&gt;   &lt;RetStart&gt;0&lt;/RetStart&gt;</span></span>
<span id="cb10-8"><a href="#cb10-8"></a><span class="co">#&gt;   &lt;IdList&gt;</span></span>
<span id="cb10-9"><a href="#cb10-9"></a><span class="co">#&gt;     &lt;Id&gt;519865230&lt;/Id&gt;</span></span>
<span id="cb10-10"><a href="#cb10-10"></a><span class="co">#&gt;     &lt;Id&gt;410810883&lt;/Id&gt;</span></span>
<span id="cb10-11"><a href="#cb10-11"></a><span class="co">#&gt;     &lt;Id&gt;313847556&lt;/Id&gt;</span></span>
<span id="cb10-12"><a href="#cb10-12"></a><span class="co">#&gt;     &lt;Id&gt;532821947&lt;/Id&gt;</span></span>
<span id="cb10-13"><a href="#cb10-13"></a><span class="co">#&gt;     &lt;Id&gt;519796743&lt;/Id&gt;</span></span>
<span id="cb10-14"><a href="#cb10-14"></a><span class="co">#&gt;     &lt;Id&gt;532821218&lt;/Id&gt;</span></span>
<span id="cb10-15"><a href="#cb10-15"></a><span class="co">#&gt;     &lt;Id&gt;519794601&lt;/Id&gt;</span></span>
<span id="cb10-16"><a href="#cb10-16"></a><span class="co">#&gt;   &lt;/IdList&gt;</span></span>
<span id="cb10-17"><a href="#cb10-17"></a><span class="co">#&gt;   &lt;TranslationSet&gt;</span></span>
<span id="cb10-18"><a href="#cb10-18"></a><span class="co">#&gt;     &lt;Translation&gt;</span></span>
<span id="cb10-19"><a href="#cb10-19"></a><span class="co">#&gt;       &lt;From&gt;Chlamydiaceae[orgn]&lt;/From&gt;</span></span>
<span id="cb10-20"><a href="#cb10-20"></a><span class="co">#&gt;       &lt;To&gt;&quot;Chlamydiaceae&quot;[Organism]&lt;/To&gt;</span></span>
<span id="cb10-21"><a href="#cb10-21"></a><span class="co">#&gt;     &lt;/Translation&gt;</span></span>
<span id="cb10-22"><a href="#cb10-22"></a>...</span></code></pre></div>
<p>Or extract parts of the XML data using the reference class method <code>#xmlValue()</code> and an XPath expression:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>cpaf<span class="op">$</span><span class="kw">xmlValue</span>(<span class="st">&quot;//Id&quot;</span>)</span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="co">#&gt; [1] &quot;519865230&quot; &quot;410810883&quot; &quot;313847556&quot; &quot;532821947&quot; &quot;519796743&quot; &quot;532821218&quot;</span></span>
<span id="cb11-3"><a href="#cb11-3"></a><span class="co">#&gt; [7] &quot;519794601&quot;</span></span></code></pre></div>
</div>
<div id="esummary-retrieving-summaries-from-primary-ids" class="section level3">
<h3><code>esummary</code>: Retrieving summaries from primary IDs</h3>
<p><code>esummary</code> retrieves document summaries (<em>docsum</em>s) from a list of primary IDs. Let’s find out what the first entry for PMP is about:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>esum &lt;-<span class="st"> </span><span class="kw">esummary</span>(cpaf[<span class="dv">1</span>])</span>
<span id="cb12-2"><a href="#cb12-2"></a>esum</span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="co">#&gt; Object of class &#39;esummary&#39; </span></span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="co">#&gt; &lt;?xml version=&quot;1.0&quot; encoding=&quot;UTF-8&quot;?&gt;</span></span>
<span id="cb12-5"><a href="#cb12-5"></a><span class="co">#&gt; &lt;!DOCTYPE eSummaryResult PUBLIC &quot;-//NLM//DTD esummary nuccore 20190808//EN&quot; &quot;https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20190808/esummary_nuccore.dtd&quot;&gt;</span></span>
<span id="cb12-6"><a href="#cb12-6"></a><span class="co">#&gt; &lt;eSummaryResult&gt;</span></span>
<span id="cb12-7"><a href="#cb12-7"></a><span class="co">#&gt;   &lt;DocumentSummarySet status=&quot;OK&quot;&gt;</span></span>
<span id="cb12-8"><a href="#cb12-8"></a><span class="co">#&gt;     &lt;DocumentSummary uid=&quot;519865230&quot;&gt;</span></span>
<span id="cb12-9"><a href="#cb12-9"></a><span class="co">#&gt;       &lt;Caption&gt;KE360863&lt;/Caption&gt;</span></span>
<span id="cb12-10"><a href="#cb12-10"></a><span class="co">#&gt;       &lt;Title&gt;Chlamydia psittaci 84-8471/1 genomic scaffold CP_84_8471_1.contig.1261, whole genome shotgun sequence&lt;/Title&gt;</span></span>
<span id="cb12-11"><a href="#cb12-11"></a><span class="co">#&gt;       &lt;Extra&gt;gi|519865230|gnl|WGS:ATMY01|CP_84_8471_1.contig.1261|gb|KE360863.1|&lt;/Extra&gt;</span></span>
<span id="cb12-12"><a href="#cb12-12"></a><span class="co">#&gt;       &lt;Gi&gt;519865230&lt;/Gi&gt;</span></span>
<span id="cb12-13"><a href="#cb12-13"></a><span class="co">#&gt;       &lt;CreateDate&gt;2013/07/02&lt;/CreateDate&gt;</span></span>
<span id="cb12-14"><a href="#cb12-14"></a><span class="co">#&gt;       &lt;UpdateDate&gt;2015/03/22&lt;/UpdateDate&gt;</span></span>
<span id="cb12-15"><a href="#cb12-15"></a><span class="co">#&gt;       &lt;Flags&gt;32&lt;/Flags&gt;</span></span>
<span id="cb12-16"><a href="#cb12-16"></a><span class="co">#&gt;       &lt;TaxId&gt;1112230&lt;/TaxId&gt;</span></span>
<span id="cb12-17"><a href="#cb12-17"></a><span class="co">#&gt;       &lt;Slen&gt;255&lt;/Slen&gt;</span></span>
<span id="cb12-18"><a href="#cb12-18"></a><span class="co">#&gt;       &lt;Biomol&gt;genomic&lt;/Biomol&gt;</span></span>
<span id="cb12-19"><a href="#cb12-19"></a><span class="co">#&gt;       &lt;MolType&gt;dna&lt;/MolType&gt;</span></span>
<span id="cb12-20"><a href="#cb12-20"></a><span class="co">#&gt;       &lt;Topology&gt;linear&lt;/Topology&gt;</span></span>
<span id="cb12-21"><a href="#cb12-21"></a><span class="co">#&gt;       &lt;SourceDb&gt;insd&lt;/SourceDb&gt;</span></span>
<span id="cb12-22"><a href="#cb12-22"></a><span class="co">#&gt;       &lt;SegSetSize&gt;0&lt;/SegSetSize&gt;</span></span>
<span id="cb12-23"><a href="#cb12-23"></a>...</span></code></pre></div>
<p>We can also parse <em>docsum</em>s into a <code>tibble</code></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>esum &lt;-<span class="st"> </span><span class="kw">esummary</span>(cpaf[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>])</span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="co">#&gt; Warning: HTTPS error: Status 429;</span></span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="kw">content</span>(esum, <span class="st">&quot;parsed&quot;</span>)</span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="co">#&gt; Warning: Errors parsing DocumentSummary</span></span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="co">#&gt; list()</span></span></code></pre></div>
</div>
<div id="efetch-downloading-full-records-from-entrez" class="section level3">
<h3><code>efetch</code>: Downloading full records from Entrez</h3>
<p>First we search the protein database for sequences of the <strong>c</strong>hlamydial <strong>p</strong>rotease <strong>a</strong>ctivity <strong>f</strong>actor, <a href="http://dx.doi.org/10.1016/j.tim.2009.07.007">CPAF</a></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>cpaf &lt;-<span class="st"> </span><span class="kw">esearch</span>(<span class="st">&quot;Chlamydia[orgn] and CPAF&quot;</span>, <span class="st">&quot;protein&quot;</span>)</span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="co">#&gt; Warning: HTTPS error: Status 429;</span></span>
<span id="cb14-3"><a href="#cb14-3"></a>cpaf</span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="co">#&gt; Object of class &#39;esearch&#39; </span></span>
<span id="cb14-5"><a href="#cb14-5"></a><span class="co">#&gt; [1] &quot;HTTPS error: Status 429; &quot;</span></span></code></pre></div>
<p>Let’s fetch the FASTA record for the first protein. To do that, we have to set <code>rettype = &quot;fasta&quot;</code> and <code>retmode = &quot;text&quot;</code>:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>cpaff &lt;-<span class="st"> </span><span class="kw">efetch</span>(cpaf[<span class="dv">1</span>], <span class="dt">db =</span> <span class="st">&quot;protein&quot;</span>, <span class="dt">rettype =</span> <span class="st">&quot;fasta&quot;</span>, <span class="dt">retmode =</span> <span class="st">&quot;text&quot;</span>)</span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="co">#&gt; Warning: HTTPS error: Status 400;</span></span>
<span id="cb15-3"><a href="#cb15-3"></a>cpaff</span>
<span id="cb15-4"><a href="#cb15-4"></a><span class="co">#&gt; Object of class &#39;efetch&#39; </span></span>
<span id="cb15-5"><a href="#cb15-5"></a><span class="co">#&gt; [1] &quot;HTTPS error: Status 400; &quot;</span></span>
<span id="cb15-6"><a href="#cb15-6"></a><span class="co">#&gt; EFetch query using the &#39;protein&#39; database.</span></span>
<span id="cb15-7"><a href="#cb15-7"></a><span class="co">#&gt; Query url: &#39;https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?=efe...&#39;</span></span>
<span id="cb15-8"><a href="#cb15-8"></a><span class="co">#&gt; Retrieval type: &#39;fasta&#39;, retrieval mode: &#39;text&#39;</span></span></code></pre></div>
<p>Now we can write the sequence to a fasta file by first extracting the data from the <code>efetch</code> object using <code>content()</code>:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="kw">write</span>(<span class="kw">content</span>(cpaff), <span class="dt">file =</span> <span class="st">&quot;~/cpaf.fna&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>cpafx &lt;-<span class="st"> </span><span class="kw">efetch</span>(cpaf, <span class="dt">db =</span> <span class="st">&quot;protein&quot;</span>, <span class="dt">rettype =</span> <span class="st">&quot;fasta&quot;</span>, <span class="dt">retmode =</span> <span class="st">&quot;xml&quot;</span>)</span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="co">#&gt; Warning: HTTPS error: Status 400;</span></span>
<span id="cb17-3"><a href="#cb17-3"></a>cpafx</span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="co">#&gt; Object of class &#39;efetch&#39; </span></span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="co">#&gt; [1] &quot;HTTPS error: Status 400; &quot;</span></span>
<span id="cb17-6"><a href="#cb17-6"></a><span class="co">#&gt; EFetch query using the &#39;protein&#39; database.</span></span>
<span id="cb17-7"><a href="#cb17-7"></a><span class="co">#&gt; Query url: &#39;https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?=efe...&#39;</span></span>
<span id="cb17-8"><a href="#cb17-8"></a><span class="co">#&gt; Retrieval type: &#39;fasta&#39;, retrieval mode: &#39;xml&#39;</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>aa &lt;-<span class="st"> </span>cpafx<span class="op">$</span><span class="kw">xmlValue</span>(<span class="st">&quot;//TSeq_sequence&quot;</span>)</span>
<span id="cb18-2"><a href="#cb18-2"></a>aa</span>
<span id="cb18-3"><a href="#cb18-3"></a><span class="co">#&gt; [1] NA</span></span>
<span id="cb18-4"><a href="#cb18-4"></a>defline &lt;-<span class="st"> </span>cpafx<span class="op">$</span><span class="kw">xmlValue</span>(<span class="st">&quot;//TSeq_defline&quot;</span>)</span>
<span id="cb18-5"><a href="#cb18-5"></a>defline</span>
<span id="cb18-6"><a href="#cb18-6"></a><span class="co">#&gt; [1] NA</span></span></code></pre></div>
</div>
<div id="einfo-information-about-the-entrez-databases" class="section level3">
<h3><code>einfo</code>: Information about the Entrez databases</h3>
<p>You can use <code>einfo</code> to obtain a list of all database names accessible through the Entrez utilities:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="kw">einfo</span>()</span>
<span id="cb19-2"><a href="#cb19-2"></a><span class="co">#&gt; Object of class &#39;einfo&#39; </span></span>
<span id="cb19-3"><a href="#cb19-3"></a><span class="co">#&gt; List of Entrez databases</span></span>
<span id="cb19-4"><a href="#cb19-4"></a><span class="co">#&gt;  [1] &quot;pubmed&quot;          &quot;protein&quot;         &quot;nuccore&quot;         &quot;ipg&quot;            </span></span>
<span id="cb19-5"><a href="#cb19-5"></a><span class="co">#&gt;  [5] &quot;nucleotide&quot;      &quot;structure&quot;       &quot;sparcle&quot;         &quot;genome&quot;         </span></span>
<span id="cb19-6"><a href="#cb19-6"></a><span class="co">#&gt;  [9] &quot;annotinfo&quot;       &quot;assembly&quot;        &quot;bioproject&quot;      &quot;biosample&quot;      </span></span>
<span id="cb19-7"><a href="#cb19-7"></a><span class="co">#&gt; [13] &quot;blastdbinfo&quot;     &quot;books&quot;           &quot;cdd&quot;             &quot;clinvar&quot;        </span></span>
<span id="cb19-8"><a href="#cb19-8"></a><span class="co">#&gt; [17] &quot;gap&quot;             &quot;gapplus&quot;         &quot;grasp&quot;           &quot;dbvar&quot;          </span></span>
<span id="cb19-9"><a href="#cb19-9"></a><span class="co">#&gt; [21] &quot;gene&quot;            &quot;gds&quot;             &quot;geoprofiles&quot;     &quot;homologene&quot;     </span></span>
<span id="cb19-10"><a href="#cb19-10"></a><span class="co">#&gt; [25] &quot;medgen&quot;          &quot;mesh&quot;            &quot;ncbisearch&quot;      &quot;nlmcatalog&quot;     </span></span>
<span id="cb19-11"><a href="#cb19-11"></a><span class="co">#&gt; [29] &quot;omim&quot;            &quot;orgtrack&quot;        &quot;pmc&quot;             &quot;popset&quot;         </span></span>
<span id="cb19-12"><a href="#cb19-12"></a><span class="co">#&gt; [33] &quot;proteinclusters&quot; &quot;pcassay&quot;         &quot;biosystems&quot;      &quot;pccompound&quot;     </span></span>
<span id="cb19-13"><a href="#cb19-13"></a><span class="co">#&gt; [37] &quot;pcsubstance&quot;     &quot;seqannot&quot;        &quot;snp&quot;             &quot;sra&quot;            </span></span>
<span id="cb19-14"><a href="#cb19-14"></a><span class="co">#&gt; [41] &quot;taxonomy&quot;        &quot;biocollections&quot;  &quot;gtr&quot;</span></span></code></pre></div>
<p>For each of these databases, we can use <code>einfo</code> again to obtain more information:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="kw">einfo</span>(<span class="st">&quot;taxonomy&quot;</span>)</span>
<span id="cb20-2"><a href="#cb20-2"></a><span class="co">#&gt; Warning: HTTPS error: Status 429;</span></span>
<span id="cb20-3"><a href="#cb20-3"></a><span class="co">#&gt; Object of class &#39;einfo&#39; </span></span>
<span id="cb20-4"><a href="#cb20-4"></a><span class="co">#&gt; [1] &quot;HTTPS error: Status 429; &quot;</span></span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
